packages/node/src/indexer/bestBlock.service.ts

```diff
@@ -0,0 +1,186 @@
+      this.finalizedBlock.block.header.number.toNumber();
+  }
+
+  storeBestBlock(blockNumber: number, hash: HexString) {
+    this.bestBlocks[blockNumber] = hash;
+  }
+  private async saveBestBlocks(
+    bestBlocks: Record<number, HexString>,
+    tx: Transaction,
+  ): Promise<void> {
+    assert(this.metaDataRepo, `Model _metadata does not exist`);
+    await this.metaDataRepo.upsert(
+      { key: METADATA_BESTBLOCKS_KEY, value: JSON.stringify(bestBlocks) },
+      { transaction: tx },
+    );
+  }
+
+  private async saveLastFinalizedVerifiedHeight(
+    height: number,
+    tx: Transaction,
+  ) {
+    assert(this.metaDataRepo, `Model _metadata does not exist`);
+    await this.metaDataRepo.upsert(
+      { key: METADATA_LAST_FINALIZED_PROCESSED_KEY, value: height },
+      { transaction: tx },
+    );
+  }
+
+  // remove any records less and equal than input finalized blockHeight
+  removeFinalized(blockHeight: number): void {
+    Object.entries(this.bestBlocks).map(([bestBlockHeight, hash]) => {
+      if (Number(bestBlockHeight) <= blockHeight) {
+        delete this.bestBlocks[bestBlockHeight];
+      }
+    });
+  }
+
+  // find closest record from block heights
+  getClosestRecord(
+    blockHeight: number,
+  ): { blockHeight: number; hash: HexString } | undefined {
+    // Have the block in the best block, can be verified
+    if (Object.keys(this.bestBlocks).length !== 0) {
+      const record = Object.entries(this.bestBlocks)
+        .sort(([bestBlockA], [bestBlockB]) => {
+          return Number(bestBlockB) - Number(bestBlockA);
+        })
+        .find(
+          ([bestBlockHeight, hash]) => Number(bestBlockHeight) <= blockHeight,
+        );
+      if (record) {
+        const [bestBlockHeight, hash] = record;
+        return { blockHeight: Number(bestBlockHeight), hash: hash };
+      }
+    }
+    return undefined;
+  }
+
+  // verify best blocks with finalized block
+  async validateBestBlocks(): Promise<boolean> {
+    const finalizedBlockNumber =
+      this.finalizedBlock.block.header.number.toNumber();
+    const lastVerifiableBlock = this.getClosestRecord(finalizedBlockNumber);
+    if (lastVerifiableBlock) {
+      if (lastVerifiableBlock.blockHeight === finalizedBlockNumber) {
+        if (lastVerifiableBlock.hash !== this.finalizedBlock.hash.toHex()) {
+          return false;
+        }
+      } else {
+        // use api to fetch lastVerifiableBlock, see if hash matches
+        const actualHash = (
+          await this.api.rpc.chain.getBlockHash(lastVerifiableBlock.blockHeight)
+        ).toHex();
+        if (actualHash !== lastVerifiableBlock.hash) {
+          logger.warn(
+            `Block folk found, enqueued un-finalized block at ${lastVerifiableBlock.blockHeight} with hash ${lastVerifiableBlock.hash}, actual hash is ${actualHash} `,
+          );
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  async getLastCorrectBestBlock(): Promise<number | undefined> {
```

Comment:
Better to use the received finalized block's parent hash and iterate backwards. `getBlockHash` might not be reliable (the node is not catching finalized block or hit cache)